CorsixTH/Lua/endconditions.lua

```diff
@@ -0,0 +1,285 @@
+    met_count, total_count = 0, 0
+    -- Look for a specific test for this criteria, or use the generic attribute test
+    local test = WinCriteria[crit_name] or WinCriteria.generic
+    gap = test(hospital, crit_name, crit_table)
+    if gap then
+      met_count = met_count + 1
+    end
+    total_count = total_count + 1
+  end
+
+  if findbest then
+    if met_count > 0 and total_count > 0 then
+      return met_count / total_count
+    else return 0
+    end
+  -- Have all criteria of the group have been met?
+  elseif met_count == total_count then
+    return "win"
+  end
+end
+
+-- Find the group of win conditions best met by the hospital.
+--!param hospital (Hospital) The hospital of the tests.
+--!return best (number) The number of the best group.
+function EndConditions:findBestWinGroup(hospital)
+  local score, best = 0, 1
+  for group = 1, self.highest_group do
+    if self.win_goals[group] then
+      local test = self:checkWinGroup(hospital, self.win_goals[group], true)
+      if test and test > score then best = group end
+    end
+  end
+  -- Return the group number of the group that meets its total most
+  return best
+end
+
+
+class "LoseCriteria"
+---@type LoseCriteria
+-- Lose criteria are met by the attribute exceeding the lose value.
+-- The generic test is used for tests on hospital attributes with the same name as the goal,
+--  eg value is hospital.value
+-- Other tests can be implemented differently.
+
+--!param hospital (Hospital) The hospital of the tests.
+--!param attribute (string) The attribute of the hospital in this test.
+--!param table (table) the lose_criteria table for one group and one criteria.
+-- Value is the hospital value.
+--!return If report is false, return the gap from hospital value to lose value, and the lose value
+-- if true, return the gap from the hospital value to boundary, and the boundary.
+function LoseCriteria.generic(hospital, attribute, table, report)
+  local target = report and table.boundary or table.lose_value
+  local measure = hospital[attribute]
+  local max_min = table.max_min_lose == 1 and 1 or -1
+  if (measure - target) * max_min > 0 then
+    return math.abs(measure - target), target
+  end
+end
+
+
+class "WinCriteria"
+--@type WinCriteria
+-- Win criteria are met by the attribute meeting or exceeding the win value.
+-- The generic test is used for tests on hospital attributes with the same name as the goal,
+--  eg value is hospital.value
+-- Other tests can be implemented differently.
+
+--!param hospital (Hospital) The hospital of the tests.
+--!param crit_name (string) The attribute of the hospital in this test.
+--!param table (table) the win_criteria table for one group and one criteria
+--!return If report is false, return the gap from hospital value to win value, and the win value.
+-- if true, return the gap from the hospital value to boundary, and the boundary.
+function WinCriteria.generic(hospital, crit_name, table, report)
```

Comment:
Does this work? I am missing the `local WinCriteria = _G["WinCriteria"]` line that makes the `WinCriteria` used here available.

Reply:
It does, even the EndCondition one is optional. I've added them.